clippy_lints/src/arc_new_in_vec_from_elem.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::higher::VecArgs;
+use clippy_utils::last_path_segment;
+use clippy_utils::macros::{root_macro_call_first_node, MacroCall};
+use rustc_hir::{Expr, ExprKind, QPath, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Arc::new` in `vec![elem; len]`
+    ///
+    /// ### Why is this bad?
+    /// This will create `elem` once and clone it `len` times - doing so with `Arc`
+    /// is a bit misleading, as it will create references to the same pointer, rather
+    /// than different `Arc` instances.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let v = vec![std::sync::Arc::new("some data".to_string()); 100];
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let data = std::sync::Arc::new("some data".to_string());
+    /// let v = vec![data; 100];
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub ARC_NEW_IN_VEC_FROM_ELEM,
+    suspicious,
+    "calling `Arc::new` in `vec![elem; len]`"
+}
+declare_lint_pass!(ArcNewInVecFromSlice => [ARC_NEW_IN_VEC_FROM_ELEM]);
+
+impl LateLintPass<'_> for ArcNewInVecFromSlice {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        let Some(macro_call) = root_macro_call_first_node(cx, expr) else { return; };
+
+        if let Some(VecArgs::Repeat(elem, _)) = VecArgs::hir(cx, expr) {
+            if !is_arc_new(cx, elem) {
+                return;
+            }
+
+            yield_lint(cx, &macro_call);
+        }
+    }
+}
+
+fn yield_lint(cx: &LateContext<'_>, macro_call: &MacroCall) {
+    span_lint_and_help(
+        cx,
+        ARC_NEW_IN_VEC_FROM_ELEM,
+        macro_call.span,
+        "calling `Arc::new` in `vec![elem; len]`",
+        None,
+        "consider extracting `Arc` initialization to a variable",
+    );
+}
+
+/// Checks whether the given `expr` is a call to `Arc::new`
+fn is_arc_new(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Call(func, _args) = expr.kind;
+        if let ExprKind::Path(ref func_path @ QPath::TypeRelative(ty, _)) = func.kind;
+        if let TyKind::Path(ref ty_path) = ty.kind;
+        if let Some(def_id) = cx.qpath_res(ty_path, ty.hir_id).opt_def_id();
+        if cx.tcx.is_diagnostic_item(sym::Arc, def_id);
+
+        then {
+            let ty_segment = last_path_segment(ty_path);
+            let func_segment = last_path_segment(func_path);
+
+            return ty_segment.ident.name == sym::Arc && func_segment.ident.name == sym::new;
+        }
```

Comment:
```suggestion
        then {
            let func_segment = last_path_segment(func_path);

            return func_segment.ident.name == sym::new;
        }
```

Reply:
> Couple more things and it should be good slightly_smiling_face


Thanks for the patience 🙂 im just getting used to the clippy concepts and api, sorry for being a bit quirky haha